### PR TITLE
Refactor connection to the backends and introduce separate classes fo…

### PR DESCRIPTION
**Pull request by vepadulano**
…r each one

This first commit introduces the changes needed to remove the old logic of a global backend running the analysis (activated through PyRDF.use)
Instead now the single dataframe instance can be connected to the backend directly. To this end a new class (PyRDF.DataFrame.SparkDataFrame) is introduced as the first one in the list of factory classes that will allow the new logic to work.
Also a factory class called DataFrameFactory has been created, allowing to dispatch the creation of a dataframe to the correct PyRDF.backend.Dist subclass. For now it mostly serves as a helper in the logic for the distributed snapshot.
Other global variables have been removed, such as include_headers and include_shared_libs. These sets of paths now are tied to the instance of the backend.

Lot of tests have been modified according to the changes, many removed. This PR takes for granted the changes introduced by #109 , so it's best to merge that one before this one.

### CODE REVIEW

File: PyRDF/DataFrame.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+import ROOT
+
+from PyRDF import Node, Proxy
+from PyRDF.backend import Spark
+
+logger = logging.getLogger(__name__)
+
+
+class SparkDataFrame(object):
```
Comment (etejedor) [inline PyRDF/DataFrame.py]: The problem with this is that we are introducing backend-dependent code in a part of the PyRDF code that should be agnostic to that.

I wonder if we could do this differently. In `PyRDF/backends` lives everything that has to do with backends. We could import whatever is in `PyRDF/backends` when importing PyRDF, and those imports would dynamically inject in PyRDF the calls that allow to create RDFs for their backends. We do a similar thing for pythonizations in ROOT:

https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/python/ROOT/__init__.py#L67-L69

The `ROOT` module does not know which pythonizations exist, but it knows they are in a subfolder `pythonization`. By importing the modules in `pythonization`, the available pythonizors for each class are registered. If a new pythonizor is added to that folder, it is automatically taken by ROOT.

Similarly, we could follow a convention so that the backends do an injection in their `__init__.py` file of their factory function. If a new backend is added, we would not need to change backend-independent code.

Comment (vepadulano) [reply PyRDF/DataFrame.py]: Interesting, looking into this :+1:

File: PyRDF/Node.py [hunk 2]
```diff
@@ -171,3 +176,176 @@ def graph_prune(self):
...
+    Attributes:
+        args (list): A list of arguments that were provided to construct
+            the RDataFrame object.
+
+
+    PyRDF's RDataFrame constructor accepts the same arguments as the ROOT's
+    RDataFrame constructor (see
+    `RDataFrame <https://root.cern/doc/master/classROOT_1_1RDataFrame.html>`_)
+
+    In addition, PyRDF allows you to use Python lists in place of C++ vectors
```
Comment (etejedor) [inline PyRDF/Node.py]: This is no longer an exclusive feature of PyRDF since new cppyy has a converter Python list -> std::vector. Anyway, we are no longer going to construct RDataFrames like this anymore, but just via the factory functions?

File: PyRDF/Node.py [hunk 3]
```diff
@@ -171,3 +176,176 @@ def graph_prune(self):
...
+                RDataFrame object.
+        """
+        super(HeadNode, self).__init__(None, None, *args)
+
+        args = list(args)  # Make args mutable
+        num_params = len(args)
+
+        for i in range(num_params):
+            # Convert Python list to ROOT CPP vector
+            if isinstance(args[i], list):
```
Comment (etejedor) [inline PyRDF/Node.py]: This is no longer necessary, see above.

File: PyRDF/Node.py (hunk 2 above)
Comment (etejedor) [reply PyRDF/Node.py]: Or is this intended to be used by the backends?

Comment (vepadulano) [reply PyRDF/Node.py]: I remember we discussed this last year and decided to keep it for backward compatibility. Since the plan now is to integrate in ROOT, anything that is the default in master today can be taken for granted I believe. In this case I can remove this logic from PyRDF.
Does this sound good ?

Comment (etejedor) [reply PyRDF/Node.py]: Yes I would remove this.

Comment (etejedor) [reply PyRDF/Node.py]: Who is intended to use this constructor on the other hand? The backends?

Comment (vepadulano) [reply PyRDF/Node.py]: Whomever will create any PyRDF.WhateverBackendDataFrame . Also the backends in the sense that `mapper` function creates ROOT.RDataFrame objects on the nodes, so this constructor will also be involved there

Comment (etejedor) [reply PyRDF/Node.py]: Ok so it's the backends, never the user code directly (PyRDF.WhateverBackendDF is also backend).

File: PyRDF/Node.py [hunk 4]
```diff
@@ -171,3 +176,176 @@ def graph_prune(self):
...
+            # If the second argument is a string
+            chain.Add(second_arg)
+        else:
+            # If the second argument is a list or vector
+            for fname in second_arg:
+                chain.Add(str(fname))
+
+        return chain.GetEntries()
+
+    def get_treename(self):
```
Comment (etejedor) [inline PyRDF/Node.py:273]: What about this API? Is this intended to be invoked by the user?

Comment (vepadulano) [reply PyRDF/Node.py:273]: That's a good question. I think it has never been discussed.

Generally, all the `get_*` functions in `HeadNode` class are used internally in PyRDF to retrieve information that was presented by the user when constructing the dataframe, e.g. the name of the TTree or the filelist or the number of entries etc. . So I would say this should not be user facing

File: PyRDF/backend/Dist.py [hunk 5]
```diff
@@ -669,7 +667,9 @@ def reducer(mergeables_out, mergeables_in):
                 for filename in value:
                     snapshot_chain.Add(filename)
                 # Create a new rdf with the chain and return that to user
-                snapshot_rdf = PyRDF.RDataFrame(snapshot_chain)
+                factory = PyRDF.DataFrame.DataFrameFactory()
```
Comment (etejedor) [inline PyRDF/backend/Dist.py]: I guess this might change depending on the changes suggested at the beginning, regarding the relocation of backend-dependent code.

File: PyRDF/backend/Dist.py [hunk 6]
```diff
@@ -685,9 +685,95 @@ def ProcessAndMerge(self, mapper, reducer):
         pass
 
     @abstractmethod
-    def distribute_files(self, includes_list):
+    def _distribute_unique_paths(self, paths):
```
Comment (etejedor) [inline PyRDF/backend/Dist.py]: What does this method do? If it is intended to be implemented by the backends, shouldn't it be not flagged as internal?

File: PyRDF/backend/Utils.py [hunk 7]
```diff
@@ -1,71 +1,135 @@
...
-            # Get return value for loading the shared library.
-            # On succesful load the value will be 0.
-            # If the library does not exist or there was an error
-            # while loading, the value will be -1
-            lib_load_return = ROOT.gSystem.Load(shared_library)
-            if lib_load_return == -1:
-                if not os.path.exists(shared_library):
-                    raise IOError("Shared library does not exist!")
-                raise Exception("ROOT couldn't load the shared library!")
+def extend_include_path(include_path):
```
Comment (etejedor) [inline PyRDF/backend/Utils.py]: Ok good, this means we won't have `Utils.Utils` anymore!

File: PyRDF/Node.py (hunk 4 above)
Comment (etejedor) [reply PyRDF/Node.py:273]: Ok, then this class should not be returned to the user. Ideally, the user should see a distributed RDataFrame as having the same API (or a subset, at least) as the plain RDataFrame, but not introduce more methods (especially if there is something already equivalent in plain RDF).

File: PyRDF/backend/Utils.py (hunk 7 above)
Comment (vepadulano) [reply PyRDF/backend/Utils.py]: This is an example of what can be done to avoid that bad pattern. 

The general rule in Python (PEP8) is to have modules name with only lowercase letters and CamelCase for classes, so in this case it could have been called `PyRDF.backend.utils.Utils` (still not ideal imho).

Since we're bound to ROOT, namespaces/modules follow the CamelCase scheme too, so our only options remain:
1. Don't put everything in a class and let free functions in the global scope of the module (like here)
2. Call classes differently from the module itself, e.g. `Utils.ClassThatKeepsUtilityFunctions`

File: PyRDF/backend/Dist.py (hunk 6 above)
Comment (vepadulano) [reply PyRDF/backend/Dist.py]: Here I'm a bit confused about the guidelines for inheritance. It is true that this method is not used directly by the `Dist` class, but at the same time it's "private" because it shouldn't be part of the user-facing API. Maybe it doesn't make a difference and I just misinterpreted the guidelines

File: PyRDF/Node.py (hunk 4 above)
Comment (vepadulano) [reply PyRDF/Node.py:273]: Ok let me understand this better. What I see when I construct a plain RDataFrame in PyROOT is
```
>>> import ROOT
>>> df = ROOT.RDataFrame(100)
>>> df.
df.Aggregate(              df.Define(                 df.GetColumnNames(         df.Histo2D(                df.Report(
df.Alias(                  df.DefineImpl(             df.GetColumnType(          df.Histo3D(                df.Snapshot(
df.AsNumpy(                df.DefineSlot(             df.GetDefinedColumnNames(  df.Max(                    df.SnapshotImpl(
df.Book(                   df.DefineSlotEntry(        df.GetFilterNames(         df.Mean(                   df.Stats(
df.Cache(                  df.Display(                df.GetNRuns(               df.Min(                    df.StdDev(
df.CacheImpl(              df.Fill(                   df.GetNSlots(              df.Profile1D(              df.Sum(
df.CheckAndFillDSColumns(  df.Filter(                 df.Graph(                  df.Profile2D(              df.Take(
df.Count(                  df.Foreach(                df.HasColumn(              df.Range(                  df.operator RInterface(
df.CreateAction(           df.ForeachSlot(            df.Histo1D(                df.Reduce(       
```

I'm not sure recreating this is possible, I would need to investigate. But here we're not giving a RDataFrame to the user, we're giving them SparkDataFrame or something similar. So generally it's not clear to me why the new distributed-aware dataframes should present the exact same API (or a subset of it)

Comment (vepadulano) [reply PyRDF/Node.py:273]: Especially for making this work with the backend itself it's necessary to have some methods or getter properties to retrieve said backend and work with that, isn't it?

Comment (vepadulano) [reply PyRDF/Node.py:273]: To clarify further, the API is gonna be there thanks to internal workings in PyRDF that connect the `Node` objects to their respective RDataFrame API callables, as was done so far in this package. Showing the same methods and attributes of the plain RDataFrame in the new `BackendDataFrame` classes is another thing though and that's where I'm in doubt

File: PyRDF/backend/Dist.py (hunk 6 above)
Comment (etejedor) [reply PyRDF/backend/Dist.py]: Ok I guess it's ok because it's meant to be invoked by another method in `Dist` (`distribute_files`) that the subclasses will inherit.

Comment (etejedor) [reply PyRDF/backend/Dist.py]: Still languages like Java don't allow private + abstract. Python is ok because there is no real private.

File: PyRDF/Node.py (hunk 2 above)
Comment (vepadulano) [reply PyRDF/Node.py]: Another thing here, there's a test that checks this functionality specifically
```python
def test_two_args(self):
    """Constructor with list of input files"""
    rdf_2_files = ["file1.root", "file2.root"]

    # Convert RDF files list to ROOT CPP vector
    reqd_vec = ROOT.std.vector('string')()
    for elem in rdf_2_files:
        reqd_vec.push_back(elem)

    # RDataFrame constructor with 2nd argument as string
    hn_1 = Node.HeadNode("treename", "file.root")

    # RDataFrame constructor with 2nd argument as Python list
    hn_2 = Node.HeadNode("treename", rdf_2_files)

    # RDataFrame constructor with 2nd argument as ROOT CPP Vector
    hn_3 = Node.HeadNode("treename", reqd_vec)

    self.assertArgs(hn_1.args, ["treename", "file.root"])
    self.assertArgs(hn_2.args, ["treename", reqd_vec])
    self.assertArgs(hn_3.args, ["treename", reqd_vec])
```

And this has made me wonder, is there a constructor in PyROOT that accepts `ROOT.std.vector("string")` as filelist? In that case I should maintain this in PyRDF too

File: PyRDF/Node.py (hunk 4 above)
Comment (etejedor) [reply PyRDF/Node.py:273]: I remember we actually discussed this in your presentation the other day, right? The idea is that the distributed RDFs look like plain RDFs, i.e. they have the same API (they can have less if it's not implemented). In the end we want the interface to be as similar as possible and the changes to the code to be minimal when switching backends.

File: PyRDF/Node.py (hunk 2 above)
Comment (etejedor) [reply PyRDF/Node.py]: PyROOT has now a converter for Python list -> std::vector thanks to cppyy. This means you don't need to maintain this conversion in PyRDF anymore, it is provided by the bindings.

File: PyRDF/backend/Dist.py (hunk 6 above)
Comment (vepadulano) [reply PyRDF/backend/Dist.py]: Ok so in light of these doubts I think it's better to remove the leading underscore

File: PyRDF/Node.py (hunk 4 above)
Comment (vepadulano) [reply PyRDF/Node.py:273]: Ok so I need to hide somehow the fact that the [Spark]DataFrame instance can have some references to the [Spark] backend. I didn't think about it this way, need to rethink this through

File: PyRDF/Node.py (hunk 2 above)
Comment (vepadulano) [reply PyRDF/Node.py]: Yep but I still saw some errors in the tests so I was confused. I might have found a bug in the conversion of std::string with TChain
```python
>>> import ROOT
>>> v = ROOT.std.vector("string")()
>>> for name in ["2clusters.root","4clusters.root"]:
...     v.push_back(name)
... 
>>> df = ROOT.RDataFrame("myTree",v)
>>> ch = ROOT.TChain("myTree")
>>> for name in v:
...     ch.Add(name)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: none of the 2 overloaded methods succeeded. Full details:
  int TChain::Add(TChain* chain) =>
    TypeError: could not convert argument 1
  int TChain::Add(const char* name, Long64_t nentries = TTree::kMaxEntries) =>
    TypeError: could not convert argument 1 (bad argument type for built-in operation)
>>> for name in v:
...     print(type(name))
...     print(name)
... 
<class cppyy.gbl.std.string at 0x55ef1a9f4d50>
2clusters.root
<class cppyy.gbl.std.string at 0x55ef1a9f4d50>
4clusters.root
```

File: PyRDF/Node.py (hunk 4 above)
Comment (vepadulano) [reply PyRDF/Node.py:273]: Latest commit has an implementation of this in the PyRDF.backend.Spark.SparkDataFrame class

File: PyRDF/Node.py (hunk 2 above)
Comment (etejedor) [reply PyRDF/Node.py]: The problem here is that, in new cppyy, when you iterate over an `std::vector<std::string>`, what you get back is `std::strings` - `name` is an `std::string`. But `TChain:Add` expects a `const char*`. The solution is not to use `std::vector` anymore in the test.

Comment (vepadulano) [reply PyRDF/Node.py]: Still there exists the possibility of constructing an RDataFrame with an std vector in python.

```python
>>> v = ROOT.std.vector("string")()
>>> for name in ["one.root","two.root"]:
...     v.push_back(name)
... 
>>> df = ROOT.RDataFrame("myTree", v)
>>> df
<cppyy.gbl.ROOT.RDataFrame object at 0x562ada82e910>
```
So that needs to be supported in PyRDF too. And since we make use of TChain, I believe the solution is to keep the conversion to a python string when constructing the TChain

Comment (etejedor) [reply PyRDF/Node.py]: Yes, it's ok to test the constructor of `RDataFrame` that accepts an `std::vector<std::string>`, just don't use what you get out of iterating that vector in `TChain::Add` (or do the conversion to Python str as you said).

File: PyRDF/Backend/Dist.py [hunk 8]
```diff
@@ -118,17 +119,8 @@ class Dist(Backend.Backend):
             information about any friend trees of the main ROOT.TTree
     """
 
-    def __init__(self, config={}):
```
Comment (etejedor) [inline PyRDF/Backend/Dist.py:121]: I wonder if it makes sense that both `Backend` and `Dist` exist, now that all backends are going to be distributed. We can still keep the hierarchy if we think it makes sense to make that distinction based on the content of these modules. Is it going to be now something like `BaseBackend` (with some basic functions) and `DistBackend` with some functions more specific to distribution?

File: PyRDF/Backend/Dist.py [hunk 9]
```diff
@@ -142,14 +134,16 @@ def __init__(self, config={}):
             'Aggregate'
         ]
 
-        # Remove the value of 'npartitions' from config dict
-        self.npartitions = config.pop('npartitions', None)
+        self.npartitions = None
```
Comment (etejedor) [inline PyRDF/Backend/Dist.py]: Why this change?

File: PyRDF/Backend/Dist.py [hunk 10]
```diff
@@ -142,14 +134,16 @@ def __init__(self, config={}):
...
 
         self.supported_operations = [op for op in self.supported_operations
                                      if op not in operations_not_supported]
 
         self.friend_info = FriendInfo()
 
+        self.headers_included = set()
```
Comment (etejedor) [inline PyRDF/Backend/Dist.py]: Just `self.headers`?

File: PyRDF/Backend/Dist.py [hunk 11]
```diff
@@ -142,14 +134,16 @@ def __init__(self, config={}):
...
+        self.shared_libraries_included = set()
```
Comment (etejedor) [inline PyRDF/Backend/Dist.py]: Just `self.shared_libraries`?

File: PyRDF/Backend/Dist.py (hunk 9 above)
Comment (vepadulano) [reply PyRDF/Backend/Dist.py]: We don't use the config dict anymore

Comment (vepadulano) [reply PyRDF/Backend/Dist.py]: I can default it to `0` maybe it's better

File: PyRDF/Backend/Dist.py (hunk 8 above)
Comment (vepadulano) [reply PyRDF/Backend/Dist.py:121]: I agree better to keep them separate for readability and clarity. Also the proposed names sound good

File: PyRDF/Backend/Dist.py (hunk 9 above)
Comment (etejedor) [reply PyRDF/Backend/Dist.py]: But we are still getting these options via the factory method for each backend, right? Then perhaps `npartitions` does not belong here anymore?

File: PyRDF/__init__.py [hunk 12]
```diff
@@ -2,20 +2,16 @@
...
 
-current_backend = None
+logger = logging.getLogger(__name__)
 
-includes_headers = set()  # All headers included in the analysis
-includes_shared_libraries = set()  # All shared libraries included
-includes_files = set()  # All other generic files included
 
-logger = logging.getLogger(__name__)
+def make_spark_dataframe(*args, **kwargs):
```
Comment (etejedor) [inline PyRDF/__init__.py]: And this goes away too

File: PyRDF/DataFrame.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
+from __future__ import print_function
+
+import logging
+import importlib
+
+logger = logging.getLogger(__name__)
+
+
+class DataFrameFactory(object):
```
Comment (etejedor) [inline PyRDF/DataFrame.py]: This guy should probably go

File: PyRDF/Proxy.py [hunk 14]
```diff
@@ -99,14 +102,13 @@ def GetValue(self):
...
         """
-        from PyRDF import current_backend
 
         # Creating a ROOT.TDirectory.TContext in a context manager so that
         # ROOT.gDirectory won't be changed by the event loop execution.
         with _managed_tcontext():
             if not self.proxied_node.value:  # If event-loop not triggered
                 generator = CallableGenerator(self.proxied_node.get_head())
-                current_backend.execute(generator)
+                self.current_backend.execute(generator)
```
Comment (etejedor) [inline PyRDF/Proxy.py]: So now the current backend is an attribute of the proxies? It should be an attribute of the dataframes, right? Or perhaps of the node we are proxying? We need to discuss where the info about the backend is stored.

File: PyRDF/Backend/Dist.py [hunk 15]
```diff
@@ -665,7 +659,9 @@ def reducer(mergeables_out, mergeables_in):
                 for filename in value:
                     snapshot_chain.Add(filename)
                 # Create a new rdf with the chain and return that to user
-                snapshot_rdf = PyRDF.RDataFrame(snapshot_chain)
+                factory = PyRDF.DataFrame.DataFrameFactory()
+                snapshot_rdf = factory.create(
+                    snapshot_chain, backend=str(generator.head_node.backend))
```
Comment (etejedor) [inline PyRDF/Backend/Dist.py]: Ok, I would get rid of the factory, because we only need the backend anyway (the backend is our factory). So we can assume that in every backend implementation there will be a free function called `MakeDataFrame`. Such function is actually the one we will inject in `RDF` as `MakeBackendDataFrame`, but it's also accessible via the backend. Sound good?

So here we would do `generator.head_node.backend.MakeDataFrame(snapshot_chain)`

File: PyRDF/Proxy.py (hunk 14 above)
Comment (vepadulano) [reply PyRDF/Proxy.py]: It's actually an attribute of the head node of the graph currently. I just alias it in the proxy function for convenience, but I agree it doesn't belong here

Comment (vepadulano) [reply PyRDF/Proxy.py]: I will reference the attribute of the head node directly

Comment (etejedor) [reply PyRDF/Proxy.py]: ok!

**Comment by vepadulano** [comment]
Latest commits implements the discussed comments.

**Review by etejedor — APPROVED** [review]

**Comment by vepadulano** [comment]
Merging :smile:

**Comment by vepadulano** [comment]
Fixed an issue with Python2, now merging
